rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs

```diff
@@ -97,90 +97,269 @@ impl From<(u64, Duration)> for Throughput {
+pub(crate) enum ThroughputReport {
+    /// Not enough data to draw any conclusions. This happens early in a request/response.
+    Incomplete,
+    /// The stream hasn't been polled for most of this time window.
+    NoPolling,
+    /// The stream has been waiting for most of the time window.
+    Pending,
+    /// The stream transferred this amount of throughput during the time window.
+    Transferred(Throughput),
+}
+
+const BUFFER_SIZE: usize = 10;
+
+/// Log of throughput in a request or response stream.
+///
+/// Used to determine if a configured minimum throughput is being met or not
+/// so that a request or response stream can be timed out in the event of a
+/// stall.
+///
+/// Request/response streams push data transfer or pending events to this log
+/// based on what's going on in their poll functions. The log tracks three kinds
+/// of events despite only receiving two: the third is "no polling". The poll
+/// functions cannot know when they're not being polled, so the log examines gaps
+/// in the event history to know when no polling took place.
+///
+/// The event logging is simplified down to a linear grid consisting of 10 "bins",
+/// with each bin representing 1/10th the total time window. When an event is pushed,
+/// it is either merged into the current tail bin, or all the bins are rotated
+/// left to create a new empty tail bin, and then it is merged into that one.
+#[derive(Clone, Debug)]
 pub(super) struct ThroughputLogs {
-    max_length: usize,
-    inner: VecDeque<(SystemTime, u64)>,
-    bytes_processed: u64,
+    resolution: Duration,
+    current_tail: SystemTime,
+    buffer: LogBuffer<BUFFER_SIZE>,
 }
 
 impl ThroughputLogs {
-    pub(super) fn new(max_length: usize) -> Self {
+    /// Creates a new log starting at `now` with the given `time_window`.
+    ///
+    /// Note: the `time_window` gets divided by 10 to create smaller sub-windows
+    /// to track throughput. The time window should configured to be large enough
+    /// so that these sub-windows aren't too small for network-based events.
+    /// A time window of 10ms probably won't work, but 500ms might. The default
+    /// is one second.
+    pub(super) fn new(time_window: Duration, now: SystemTime) -> Self {
+        assert!(!time_window.is_zero());
+        let resolution = time_window.div_f64(BUFFER_SIZE as f64);
         Self {
-            inner: VecDeque::with_capacity(max_length),
-            max_length,
-            bytes_processed: 0,
+            resolution,
+            current_tail: now,
+            buffer: LogBuffer::new(),
         }
     }
 
-    pub(super) fn push(&mut self, throughput: (SystemTime, u64)) {
-        // When the number of logs exceeds the max length, toss the oldest log.
-        if self.inner.len() == self.max_length {
-            self.bytes_processed -= self.inner.pop_front().map(|(_, sz)| sz).unwrap_or_default();
-        }
+    /// Returns the resolution at which events are logged at.
```

Comment:
need an example of some sort to understand what resolution means in this context. Is it the width of a bin?

Reply:
I think it is. A resolution is the amount of time each cell is responsible for.